### PR TITLE
Make int unspecialization actually work

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95975
* __->__ #95621

OK, so this PR used to be about reducing the number of constants we specialize on, but it turns out that unspecialization was ~essentially never used (because we still constant specialized way too aggressively) and I ended up having to fix a bunch of issues to actually get tests to pass. So this PR is now "make int unspecialization actually work". As part of this, I have to turn off unspecialization by default, as there are still latent bugs in inductor.

The general strategy is that an unspecialized int is represented as a SymInt. Representing it as a 0d tensor (which is what the code used to do) is untenable: (1) we often need unspecialized ints to participate in size computations, but we have no way of propagating sympy expressions through tensor compute, and (2) a lot of APIs work when passed SymInt, but not when passed a Tensor. However, I continue to represent Numpy scalars as Tensors, as they are rarely used for size computation and they have an explicit dtype, so they are more accurately modeled as 0d tensors.

* I folded in the changes from https://github.com/pytorch/pytorch/pull/95099 as I cannot represent unspecialized ints as SymInts without also turning on dynamic shapes. This also eliminates the necessity for test_unspec.py, as toggling specialization without dynamic shapes doesn't do anything. As dynamic shapes defaults to unspecializing, I just deleted this entirely; for the specialization case, I rely on regular static shape tests to catch it. (Hypothetically, we could also rerun all the tests with dynamic shapes, but WITH int/float specialization, but this seems... not that useful? I mean, I guess export wants it, but I'd kind of like our Source heuristic to improve enough that export doesn't have to toggle this either.)
* Only 0/1 integers get specialized by default now
* A hodgepodge of fixes. I'll comment on the PR about them.

Fixes https://github.com/pytorch/pytorch/issues/95469

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/config.py
Comment (bdhirsh) [inline torch/_dynamo/config.py:46]: for my understanding - why is this default False? Is this config a no-op unless `dynamic_shapes=True` is set?

Comment (ezyang) [reply torch/_dynamo/config.py:46]: So the idea is post Avik's PR, `specialize_int_float` does nothing unless dynamic_shapes=True is set, and we don't want to specialize on int inputs (see for example https://github.com/pytorch/pytorch/issues/93468#issuecomment-1442732186 of overspecialization even with dynamic=True). But export might want to specialize, so we let them turn ON specialization if they want.

File: torch/_dynamo/variables/torch.py
Comment (ezyang) [inline torch/_dynamo/variables/torch.py:561]: Right now, you can't pass SymInt to torch.tensor constructor, because make_fx tracing doesn't know how to deal with it, since we just inline tensor constant into graph. Actually, Dynamo export can deal with it just fine. This is the `int_tensor` thing.

This isn't the real reason I added this code though: originally when we treated unspec ints as 0d tensors, I would sometimes get a torch.tensor([fake_tensor]) call, where the fake tensor represented an unspecialized 0d scalar. This doesn't work, cuz we try to item() the fake tensor to get the real value out but there's no real value. However, you shouldn't really hit this case anymore (unless you make a numpy scalar and pass it to torch.tensor but I REFUSE TO ADD THIS TEST because no one actually is going to hit this in practice.)

File: torch/_dynamo/variables/builtin.py
Comment (ezyang) [inline torch/_dynamo/variables/builtin.py:610]: This is numpy scalar shenanigans. If I have max(numpy_scalar, sym_int), I MUST NOT run the Tensor operation, because torch.maximum doesn't accept int argument. The directionality here goes both ways, as is being adjusted here.

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:192]: Probably not going to specialize these, might even update on this PR

File: torch/_dynamo/utils.py
Comment (ezyang) [inline torch/_dynamo/utils.py:964]: Actually maybe this is wrong

File: torch/_dynamo/utils.py
Comment (ezyang) [inline torch/_dynamo/utils.py:967]: https://github.com/pytorch/pytorch/pull/95707

File: torch/_dynamo/variables/tensor.py
Comment (ezyang) [inline torch/_dynamo/variables/tensor.py:459]: This appears to have Consequences(TM)

File: torch/_dynamo/variables/builtin.py
Comment (ngimel) [reply torch/_dynamo/variables/builtin.py:610]: But you can run `clamp` as below, it accepts int args?

Comment (ezyang) [reply torch/_dynamo/variables/builtin.py:610]: Oh yeah, I could alternately reroute this to run clamp instead of maximum/minimum. Uhhh I'm not sure which is better.

File: test/dynamo/test_dynamic_shapes.py
Comment (ezyang) [inline test/dynamo/test_dynamic_shapes.py:125]: This one is from Avik's PR, see discussion there https://github.com/pytorch/pytorch/pull/95099

File: torch/_dynamo/config.py
Comment (ezyang) [inline torch/_dynamo/config.py:46]: I decided that I am only going to unspecialize ints, and NOT floats. This side-steps the discussion of whether or not float inputs should be repped as 0d tensors or SymFloat.

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:192]: UPDATE: These constants actually don't do anything in this version of the PR. I'll cleanup later.

File: torch/_inductor/codegen/common.py
Comment (ezyang) [inline torch/_inductor/codegen/common.py:66]: I'm not sure about this. The changes here are consistent with @nkaretnikov's changes in https://github.com/pytorch/pytorch/pull/95698 and https://github.com/pytorch/pytorch/pull/95773 but honestly this class of changes gives me the jeebies.

One thing that particularly annoys me is the conversion of base into a float32 before sqrt'ing it. This feels... bad. Why not a float64? Well, that actually doesn't work because downstream seems to expect this to be a float32. How do I even tell what the width of the type is supposed to be here. Sympy is just not giving us enough rope.

File: torch/_inductor/utils.py
Comment (ezyang) [inline torch/_inductor/utils.py:85]: I hate this. But I don't really understand how I'm getting a Tensor as argument here. The backtrace looks like:

```
  File "/data/users/ezyang/a/pytorch/torch/_inductor/compile_fx.py", line 250, in run                               return model(new_inputs)                                                                                    
  File "/tmp/torchinductor_ezyang/o4/co44acxg774qgd4loteojd3mm7pnotnp5bdhgqch3zi3kbdp43re.py", line 558, in call
    triton__5.run(buf19, buf22, s2, triton__5_xnumel, 512, grid=grid(triton__5_xnumel), stream=stream0)           File "/data/users/ezyang/a/pytorch/torch/_inductor/triton_ops/autotune.py", line 185, in run                      return launcher(                                                                                            
  File "<string>", line 3, in launcher                                                                          
  File "/data/users/ezyang/a/pytorch/torch/_inductor/triton_ops/autotune.py", line 707, in grid_fn              
    get_grid_dim(xnumel, meta.get("XBLOCK", None)),                                                             
  File "/data/users/ezyang/a/pytorch/torch/_inductor/triton_ops/autotune.py", line 703, in get_grid_dim             return ceildiv(numel, block)                                                                                
  File "/data/users/ezyang/a/pytorch/torch/_inductor/utils.py", line 81, in ceildiv                             
    assert isinstance(numer, int) and isinstance(denom, int), f"ceildiv({numer}, {denom})"                      
AssertionError: ceildiv(3072, 1)                                                                                
TorchDynamo optimized model failed to run because of following error 
```

File: torch/fx/experimental/symbolic_shapes.py
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1340]: This is not currently used; was briefly used when I was playing around with SymFloat arguments to graphs.

File: torch/utils/_sympy/interp.py
Comment (ezyang) [inline torch/utils/_sympy/interp.py:27]: CleanDiv is the one that showed up when traversing sympy expressions, CeilDiv/ModularIndexing didn't show up in testing.

File: torch/_inductor/codegen/common.py
Comment (nkaretnikov) [reply torch/_inductor/codegen/common.py:66]: This is `ExprPrinter`, `tl.libdevice.*` should be in `TritonPrinter`, no?

**Comment by ezyang** [comment]
This is very much ready for review

File: test/dynamo/test_dynamic_shapes.py
Comment (ezyang) [reply test/dynamo/test_dynamic_shapes.py:104]: What we would like to happen here is that we force a guard on the int when it slices into a Python list. But the usage of the int happens deep inside a slice, so it is not obvious at the call site that this is possible. I also tried making as_python_constant work universally for SymNodeVariable, but this causes a ton of fallout (as code now assumes that it can exercise this path.) Probably the right thing to do is add another method for this sort of thing.

File: test/dynamo/test_unspec.py
Comment (ezyang) [inline test/dynamo/test_unspec.py:50]: I've moved the unspec tests over to dynamic shapes since it's simpler to do it that way

File: torch/_dynamo/variables/builtin.py
Comment (ezyang) [reply torch/_dynamo/variables/builtin.py:610]: Given our new design in https://docs.google.com/document/d/1INSCdYu1PxXcr43HrD82OudeEuS-qxQe1yZmLg2wy6A/edit#  because numpy_scalar is repped as a 0d tensor this means that we SHOULD tensorify here. I'll do this update.

File: torch/_dynamo/utils.py
Comment (Chillee) [reply torch/_dynamo/utils.py:964]: Why?

File: torch/_dynamo/variables/builder.py
Comment (Chillee) [inline torch/_dynamo/variables/builder.py:179]: wtf was this doing

File: torch/_dynamo/variables/builder.py
Comment (Chillee) [inline torch/_dynamo/variables/builder.py:802]: I think this comment can be removed with the current plan?

**Comment by ezyang** [comment]
I did a sweep with this PR and --unspecialize-int. Here are the failures

image

Sheet at https://docs.google.com/spreadsheets/d/1-ghURoD6tzTushbCb22tFd5B5uNg7HSxVG4dQBwoaFA/edit?pli=1#gid=921702933 and logs at https://gist.github.com/c5903d1de28012b3e117b1ef67260a57

**Review by Chillee — APPROVED** [review]
This PR mostly looks pretty ... boring? Seems like it's primarily just changing some policies on where we specialize and fixing some other cases.

Am I missing something 🤔

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:179]: Specialize every number that is less than 17 lol. This is what I mean by "int unspecialization didn't work"

File: torch/_dynamo/variables/builder.py
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:802]: yes

File: torch/_dynamo/utils.py
Comment (ezyang) [reply torch/_dynamo/utils.py:964]: we only tested if ref was a numpy int type, not res... actually I also see that I didn't test float. I guess I should update this.

**Comment by ezyang** [comment]
@Chillee well, I *originally* thought this was just going to be as simple as removing the range(17) constants, but there were a bunch of knock on effects haha.

**Comment by atalman** [comment]
cc @guangy10 @ezyang 

This issue is in the  milestones :[ https://github.com/pytorch/pytorch/milestone/36?closed=1](https://github.com/pytorch/pytorch/milestone/36?closed=1), if you want to see your fix included in this minor release. Please post it as a cherry-pick into the [[v2.0.1] Release Tracker](https://github.com/pytorch/pytorch/issues/97272). 

**The deadline is April 14, 5PM PST.**

**Only issues that have ‘cherry-picks’ will be considered for the release.**

Common FAQs:

Q1: Where can I find more information on the release process and terminology?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md)

Q2: Am I guaranteed to be included in the cherry-pick if I do above?

A: No, it is not guaranteed, the Release Team will review all submissions against the listed criteria before making the final decision on what to include on 4/17.

Q3: When is 2.1 going to be released?

A: We do not have a formal date at this time but will update the community when we do. Our immediate focus is 2.0.1. Note that 1.12 was released on 6/28/22, 1.13 on 10/28/22 and 2.0 on 3/15/23.

Q4: **I missed the 4/14 5PM PST deadline, is there any option to have an extension?**

A: **No, in order to meet our 4/28 goal, we must hold 4/14 as our deadline and will not accept any requests after the fact. We are over communicating the timelines and process with the community to avoid such issues.**

Q5: Where should I double check to see if my issue is in the cherry pick tracker?

A: [[v2.0.1] Release Tracker · Issue #97272 · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/issues/97272)

Q6: Where can I find the Release Compatibility Matrix for PyTorch?

A: [pytorch/RELEASE.md at master · pytorch/pytorch · GitHub](https://github.com/pytorch/pytorch/blob/master/RELEASE.md#release-compatibility-matrix)

Please contact OSS Releng team members if you have any questions/comments. Again we appreciate everyone’s time and commitment to the community, PyTorch and 2.0 and 2.01 releases!


Please refer to this post for more details: https://dev-discuss.pytorch.org/t/pytorch-release-2-0-1-important-information/1176

**Pull request closed without merging**
